### PR TITLE
Revert changes in tuple hashing algorithm

### PR DESCRIPTION
**Pull request by jkotas**

### CODE REVIEW

**Comment by jkotas** [comment]
cc @jamesqo

File: src/mscorlib/src/System/Tuple.cs [hunk 1]
```diff
@@ -1,13 +1,16 @@
...
 // See the LICENSE file in the project root for more information.
-
 using System;
 using System.Text;
 using System.Collections;
 using System.Collections.Generic;
 using System.Diagnostics.Contracts;
 
+//
+// Note: F# compiler depends on the exact tuple hashing algorithm. Do not ever change it.
```
Comment (stephentoub) [inline src/mscorlib/src/System/Tuple.cs]: Is F# going to need the same hashing algorithm for ValueTuple?

**Comment by jamesqo** [comment]
@jkotas, maybe just the part that calls into HashHelpers (the actual hashing algorithm) and CombineHashCodes could be reverted? That way we would still avoid boxing.

**Comment by jkotas** [comment]
Yes, it may be possible to do this in non-breaking way. The F# tests should to be run on the change to verify that it does not break it again, and some data that it actually makes things better (and not worse) for F# better would be useful too.

**Comment by jkotas** [comment]
Another angle to consider - Tuple is not sealed, and so it is possible to inherit from it and override just the IStructuralEquatable method... .

**Comment by jamesqo** [comment]
@jkotas, how could it be worse if we avoid boxing?

> Tuple is not sealed, and so it is possible to just inherit from it

Most people use Tuple without inheriting from it, though.

**Comment by jkotas** [comment]
The known primary user of Tuple is F#. Boxing is not end of the world - you will find quite a bit of boxing in any real world program. F# programs tend to allocate a lot, but they still run fast because of the allocation pattern is very friendly to GC -  there are next to none references from higher generations to Gen0 that make the GC very cheap. It is the fairly typical for functional languages - the pure functional languages tend to have the GC designed around it.

> Most people use Tuple without inheriting from it, though

I can believe it in C#. But is it the case in other languages that run on .NET? It does not sound unreasonable for compiler generated code to inherit from Tuple, e.g. to give names to fields.

**Pull request merged**